### PR TITLE
fix: Forbid SSO users from local logins or password changes

### PR DESCRIPTION
**Pull request by codingllama**
Avoids possible scenarios where an SSO user gets a password for themselves and uses it to skip the actual SSO authentication.

Locking in local logins should be enough (AuthenticateSSHUser and AuthenticateWebUser), but I've gone the extra mile to forbid SSO users from having passwords or recovery codes as well.

It's worth noting that when I say "SSO user" I actually mean "non-local user".

Changelog: Forbid SSO users from local logins or password changes

https://github.com/gravitational/security-findings/issues/87

### CODE REVIEW

File: lib/auth/accountrecovery.go [hunk 1]
```diff
@@ -351,6 +354,15 @@ func (a *Server) CreateAccountRecoveryCodes(ctx context.Context, req *proto.Crea
 		return nil, trace.AccessDenied(unableToCreateCodesMsg)
 	}
 
+	// Verify if the user is local.
+	switch user, err := a.GetUser(ctx, token.GetUser(), false /* withSecrets */); {
+	case err != nil:
+		// err swallowed on purpose.
+		return nil, trace.AccessDenied(unableToCreateCodesMsg)
```
Comment (codingllama) [inline lib/auth/accountrecovery.go:361]: I've opted to be opaque here, following the surrounding code, but didn't feel the same need below.

File: lib/auth/methods.go [hunk 2]
```diff
@@ -378,6 +385,22 @@ func (a *Server) authenticateUserInternal(
 		return a.authenticatePasswordless(ctx, req)
 	}
 
+	// Disallow non-local users from local authentication.
+	// Passwordless does its own check, as the user is only known after the
+	// webauthn assertion is cleared.
+	switch u, err := a.GetUser(ctx, user, false /* withSecrets */); {
+	case trace.IsNotFound(err):
+		// Keep going if the user is not known.
```
Comment (codingllama) [inline lib/auth/methods.go:393]: This was the case before and is covered by other tests, so seemed reasonable to keep the behavior.

File: lib/auth/methods.go [hunk 3]
```diff
@@ -378,6 +385,22 @@ func (a *Server) authenticateUserInternal(
...
+	case err != nil:
+		return nil, "", trace.Wrap(err)
```
Comment (codingllama) [inline lib/auth/methods.go:395]: Arguably this could be some sort of AccessDenied or internal error.

File: lib/auth/usertoken.go [hunk 4]
```diff
@@ -148,10 +148,11 @@ func (a *Server) CreateResetPasswordToken(ctx context.Context, req CreateUserTok
 		return nil, trace.BadParameter("invalid reset password token request type")
 	}
 
-	// ResetPassword doesn't error for non-existent users, so first check if the
-	// user exists.
-	if _, err = a.GetUser(ctx, req.Name, false /* withSecrets */); err != nil {
+	switch user, err := a.GetUser(ctx, req.Name, false /* withSecrets */); {
+	case err != nil:
 		return nil, trace.Wrap(err)
```
Comment (codingllama) [inline lib/auth/usertoken.go:153]: It didn't seem to me that important to hide the error here either.

**Review by kimlisa — APPROVED** [review]

**Comment by codingllama** [comment]
I was just taking a look at the FTD. TestCreateAccountRecoveryCodes is kind of slow, TestAuthenticateSSHUser is just awful (so much that I'm tempted to revert that one line I touched there). I'll see if I can find some easy wins, but otherwise I think an FTD exemption could make more sense for this one.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**